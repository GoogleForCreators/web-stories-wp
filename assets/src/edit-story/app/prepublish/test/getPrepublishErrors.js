/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Internal dependencies
 */
import { getPrepublishErrors } from '..';
const DEFAULT_STORY_PROPS = {
  featuredMedia: {
    url: 'https://greatimageaggregate.com/1234',
  },
  title: 'How to get rich',
  excerpt: "There's a secret no one wants you to know about",
};
describe('prepublish checklist', () => {
  beforeEach(jest.resetModules);
  it('should return guidance for the story', async () => {
    const testStory = {
      id: 120,
      pages: [{ id: 123, elements: [] }],
    };
    const result = await getPrepublishErrors(testStory);
    expect(result).toMatchSnapshot();
  });
  it('should return an empty array', async () => {
    const testStory = {
      id: 456,
      ...DEFAULT_STORY_PROPS,
      pages: [
        {
          id: 1,
          elements: [
            {
              type: 'text',
              content: 'The prepublish checklist should return an empty array',
            },
          ],
        },
        {
          id: 2,
          elements: [
            {
              type: 'text',
              content: 'A story needs at least 100 characters',
            },
          ],
        },
        {
          id: 3,
          elements: [
            {
              type: 'text',
              content: 'There once was a man from Nantucket...',
            },
          ],
        },
        {
          id: 4,
          elements: [],
        },
      ],
    };
    expect((await getPrepublishErrors(testStory)) instanceof Array).toBeTrue();
    expect(await getPrepublishErrors(testStory)).toHaveLength(0);
  });

  it('should not throw errors if the checklist functions throw (and should not skip the others)', async () => {
    const mockFn = jest.fn(() => {
      throw new Error('What are you doing here?');
    });
    jest.doMock('../warning', () => {
      const actual = jest.requireActual('../warning').default;
      const async = jest.requireActual('../warning').async;
      return {
        ...actual,
        async,
        story: [...actual.story, mockFn],
        page: [...actual.page, mockFn],
        image: [...actual.image, mockFn],
      };
    });
    const getPrepublishErrorsCopy = jest.requireActual('../getPrepublishErrors')
      .default;
    const malformedStory = {
      title: undefined,
      pages: [{ elements: [{ type: 'image', height: 1, width: 1 }] }],
    };
    expect(async () => {
      const result = await getPrepublishErrorsCopy(malformedStory);
      return result;
    }).not.toThrow();
    expect(mockFn).toHaveBeenCalledWith(malformedStory);
    expect(mockFn).toHaveBeenCalledTimes(3);
    expect(await getPrepublishErrorsCopy(malformedStory)).toStrictEqual(
      await getPrepublishErrors(malformedStory)
    );
  });

  it('should provide the page number where the element that needs guidance is', async () => {
    const malformedStory = {
      title: undefined,
      pages: [
        {
          elements: [
            {
              id: 456,
              type: 'text',
              content: 'some normal text content, should not appear',
            },
          ],
        },
        {
          elements: [{ id: 123, type: 'image', height: 1, width: 1 }],
        },
      ],
    };
    const test = await getPrepublishErrors(malformedStory);
    const elementErrors = test.filter(({ elementId }) => Boolean(elementId));
    expect(elementErrors).toHaveLength(1);
    expect(elementErrors[0].page).toStrictEqual(2);
  });
});
