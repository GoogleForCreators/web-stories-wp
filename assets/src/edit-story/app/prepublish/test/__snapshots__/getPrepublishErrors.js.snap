// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepublish checklist should return guidance for the story 1`] = `
Array [
  Object {
    "help": <ul>
      <li>
        Use as a representation of the story
      </li>
      <li>
        Avoid images with embedded text
      </li>
      <li>
        Use an image that's at least 640x853px
      </li>
      <li>
        Maintain a 3:4 aspect ratio
      </li>
    </ul>,
    "highlight": "POSTER",
    "message": "Add cover image",
    "storyId": 120,
    "type": "error",
  },
  Object {
    "help": <ul>
      <li>
        Keep under 10 words
      </li>
      <li>
        Don't exceed 40 characters
      </li>
    </ul>,
    "highlight": "STORY_TITLE",
    "message": "Add story title",
    "storyId": 120,
    "type": "error",
  },
  Object {
    "help": <ul>
      <li>
        Incorporate a brief description for better user experience
      </li>
    </ul>,
    "highlight": "EXCERPT",
    "message": "Add story description",
    "storyId": 120,
    "type": "warning",
  },
  Object {
    "help": <ul>
      <li>
        It is recommended to have between 4 and 30 pages in your story
      </li>
    </ul>,
    "message": "Make story longer",
    "storyId": 120,
    "type": "guidance",
  },
  Object {
    "help": <ul>
      <li>
        Make your story text 100 characters or more
      </li>
    </ul>,
    "message": "Increase amount of total text",
    "storyId": 120,
    "type": "guidance",
  },
]
`;
